Battery attribute & accuracy filter

GoNzCiD opened the pull request:
## Closed to continue in home-assistant/home-assistant/pull/24277

## Breaking Change:
  - ```battery_level``` attribute changes to ```battery```.
Since a ```device_tracker``` component contains a standart [attribute](https://github.com/home-assistant/home-assistant/blob/master//homeassistant/components/device_tracker/__init__.py#L63) for ```battery``` level, that we can use with [see](https://www.home-assistant.io/components/device_tracker/#device_trackersee-service) service, update with [async_see](https://github.com/home-assistant/home-assistant/blob/master//homeassistant/components/device_tracker/__init__.py#L274) method, we should use it for maintain component criteria and not create a new attribute for the same pourpose with a different name.
To solve this breaking change, only need to replace the ```battery_level``` attribute with ```battery``` in all templates than use the battery attribute of any _traccar device_tracker_. 

## Description:
  - Battery attribute is sent into see method instead of an additional paremeter.
  - Added an accuracy filter option to exclude positions if accuracy is higher than specified.
  - Added a skip accuracy filter option to avoid filter certain positions by accuracy if any of the specified attributes are present in the message, ie throw an alarm but accuracy is too high.
  - Extract constants to _const.py_ file.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#9347

## Example entry for `configuration.yaml` (if applicable):
```yaml
device_tracker:
  - platform: traccar
    ...
    max_accuracy: 100
    skip_accuracy_filter_on:
      - alarm
    ...
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly ([example][ex-manifest]).
  - [x] New dependencies have been added to `requirements` in the manifest ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-manifest]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json
[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json#L5
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23
[manifest-docs]: https://developers.home-assistant.io/docs/en/development_checklist.html#_the-manifest-file_

GoNzCiD committed:
use 'battery' attribue for device_trackers instead of custom 'battery_level' attribute

GoNzCiD committed:
trailing whitespace fix

GoNzCiD committed:
extract const to const.py file

GoNzCiD committed:
Add new line at EOF

GoNzCiD committed:
add filter by max accuracy

GoNzCiD committed:
trailing whitestapce

GoNzCiD committed:
Fix static-checks

GoNzCiD committed:
more static-checks

GoNzCiD committed:
omit const.py from coverage

GoNzCiD committed:
Add option to skip accuracy filter when certain attributes are sent from traccar

ludeeus commented on homeassistant/components/traccar/const.py:
```diff
@@ -0,0 +1,31 @@
+"""Constants for Traccar platform."""
```
Drop "platform" here or change it to "integration"

GoNzCiD replied:
Fixed

ludeeus commented on homeassistant/components/traccar/device_tracker.py:
```diff
@@ -172,11 +159,24 @@ def __init__(self, api, hass, async_see, scan_interval,
                 if device_info.get(custom_attr) is not None:
                     attr[custom_attr] = device_info[custom_attr]
+                    if custom_attr in self._skip_accuracy_on:
+                        skip_accuracy_filter = True
```
Since this is part of the loop it will set the value for the last item in that loop.

GoNzCiD replied:
The goal is to set ```skip_accuracy_filter``` var to true when at least one of the items of ```_custom_attributes``` are contained on ```_skip_accuracy_on``` collection and are present in traccar message.
ie. Position has an higher accuracy but message contains alarm attribute, so it's importatn to receive despite of accuracy.

ludeeus commented on homeassistant/components/traccar/device_tracker.py:
```diff
@@ -172,11 +159,24 @@ def __init__(self, api, hass, async_see, scan_interval,
+            accuracy = float(0)
+            if device_info.get('accuracy') is not None:
+                accuracy = device_info['accuracy']
+            if (not skip_accuracy_filter and self._max_accuracy > 0 and
```
Since `skip_accuracy_filter`  are set inside a loop, and this is used outside.
This type of filtering would not give the expected result (if you have multiple custom_attrs, for 1 it will not matter).

GoNzCiD replied:
Related to previous lines comment, ```skip_accuracy_filter``` var is initialited to ```False``` on line [138](https://github.com/home-assistant/home-assistant/pull/23292/files#diff-5192f6e34a898c79e0b402348c7a7a65R138) outside the loop. It only shold be true when traccar message contains an attribute listed in configuration under ```skip_accuracy_filter_on```.
ie. This configuration should allow to have ```alarm``` and ```last_update``` under ```monitored_conditions``` (custom_attrs) but only ```alarm``` under ```skip_accuracy_filter_on```, so all traccar api reponses retrive ```last_update``` and will be filtered if accuracy is higher than limit, but it will not be filtered if contains an ```alarm```.

ludeeus replied:
I see, so the design is that if _any_ attribute is in `_skip_accuracy_on` it should enable `skip_accuracy_filter` for _all_ attributes.

GoNzCiD replied:
Yes!

ludeeus commented on homeassistant/components/traccar/device_tracker.py:
```diff
@@ -172,11 +159,24 @@ def __init__(self, api, hass, async_see, scan_interval,
+                accuracy = device_info['accuracy']
+            if (not skip_accuracy_filter and self._max_accuracy > 0 and
+                    accuracy > self._max_accuracy):
+                _LOGGER.info('Excluded position by accuracy filter: %f (%s)',
```
This would be "spammy", change it to debug or consider removing it.

GoNzCiD replied:
OK, Changed to debug.

ludeeus commented on homeassistant/components/traccar/device_tracker.py:
```diff
@@ -172,11 +159,24 @@ def __init__(self, api, hass, async_see, scan_interval,
                      device_info.get('longitude')),
-                gps_accuracy=(device_info.get('accuracy')),
+                gps_accuracy=accuracy,
+                battery=device_info.get('battery'),
```
Would this result in a breaking change for users that have set up templates to get the battery?

GoNzCiD replied:
Oh yes! this is a breaking change for these users since [160](https://github.com/home-assistant/home-assistant/pull/23292/files#diff-5192f6e34a898c79e0b402348c7a7a65L160) and [161](https://github.com/home-assistant/home-assistant/pull/23292/files#diff-5192f6e34a898c79e0b402348c7a7a65L161) lines has been deleted from actual version.
I think this is the right way to push battery info in the device_tracker, it's expected on async_see method so I don't know the reason to create a new custom attribute for the device_tracker.

With this modification device_tracker attribute named battery_level in the past now is named battery. I will document it in the PR.

ludeeus replied:
Can you add some more description to the breaking change note?:
- Why it's needed
- What the user needs to do

GoNzCiD replied:
Please, check if it's better now :)

ludeeus replied:
Much better :)

ludeeus reviewed (changes requested):
Left a few notes for you to start this off :)

GoNzCiD committed:
Fix description commentary

GoNzCiD committed:
Low log level to debug in order to avoid possible log spamming

GoNzCiD committed:
Fix indentation

pvizeli committed:
Fix Hassio-version for Azure Pipelines (#23895)

balloob committed:
Update Honeywell warning (#23913)

elupus committed:
Switch media player to SWITCH type (#23914)

MEDIA device type is being rejected by google now.

karlkar committed:
Fix problem with cameras that don't support time (#23924)

Some onvif cameras don't support Date management. In that case None is returned and script crashes when trying to obtain date

balloob committed:
Bumped version to 0.93.1

balloob committed:
Merge pull request #23932 from home-assistant/rc

0.93.1

pvizeli committed:
Fix auto version update Hass.io (#23935)

ludeeus committed:
Bump pytraccar (#23939)

aerialls committed:
Setup integration dependencies before loading it (#23957)

Danielhiversen committed:
upgrade broadlink library (#23966)

cyrosy committed:
Make Discord payload data key not required (#23964)

ludeeus committed:
Fixes issue with multiple alerts (#23945)

* Fixes issue with multiple alerts

* Adds missing new line

* Remove whitespace

balloob committed:
Bumped version to 0.93.2

balloob committed:
Merge pull request #24047 from home-assistant/rc

0.93.2

GoNzCiD committed:
Merge branch 'dev' into dev

GoNzCiD committed:
Align version with home-assistant/dev branch

GoNzCiD committed:
Merge branch 'dev' of https://github.com/GoNzCiD/home-assistant into dev

GoNzCiD commented:
Outdated => home-assistant/home-assistant/pull/24277
